external_services/lti1p3.py

```diff
@@ -0,0 +1,360 @@
+        self.user = user
+        self.redirect_uri = params.get('redirect_uri')
+        self.login_hint = params.get('login_hint')
+        self.instance = None
+        self.exercise = None
+
+        if not message_hint:
+            raise OAuth2Error("Missing message_hint")
+
+        hint = message_hint.split(':')
+        try:
+            self.instance = CourseInstance.objects.get(pk=int(hint[0]))
+        except CourseInstance.DoesNotExist as e:
+            raise OAuth2Error("Invalid message_hint: cannot find course instance") from e
+
+        if len(hint) > 1:
+            # When accessed from side menu, LTI request is not associated with exercise,
+            # and then the exercise part of the hint is not defined.
+            try:
+                self.exercise = LTI1p3Exercise.objects.get(pk=int(hint[1]))
+            except LTI1p3Exercise.DoesNotExist:
+                pass
+
+        client_id = params.get('client_id')
+        if client_id:
+            try:
+                self.service = LTI1p3Service.objects.get(client_id=client_id)
+            except LTI1p3Service.DoesNotExist as e:
+                raise OAuth2Error("Invalid client ID") from e
+        else:
+            raise OAuth2Error("Client ID is not specified")
+
+    def validate_client_id(self, client_id: str, request, *args, **kwargs) -> bool:
+        if (self.service and self.service.client_id == client_id):
+            return True
+
+        logger.error("LTI 1.3: validate client failed for '%s'", client_id)
+        return False
+
+    def validate_redirect_uri(self, client_id: str, redirect_uri: str, request, *args, **kwargs) -> bool:
+        # For now we do not have means for better validation, redirect_uri is not preconfigured.
+        if redirect_uri:
+            return True
+        logger.error("LTI 1.3: validate redirect_uri failed for client '%s', uri: %s", client_id, redirect_uri)
+        return False
+
+    def validate_response_type(self, client_id: str, response_type: str, client, request, *args, **kwargs) -> bool:
+        if (response_type == 'id_token'):
+            return True
+
+        logger.error("LTI 1.3: validate response_type failed for client '%s', type: %s", client_id, response_type)
+        return False
+
+    def validate_scopes(self,
+            client_id: str,
+            scopes: List[str],
+            client: LTI1p3Client,
+            request,
+            *args,
+            **kwargs,
+            ) -> bool:
+        # Auth login request, should always be 'openid'
+        if scopes[0] != "openid":
+            logger.error("LTI 1.3: Invalid scope for login request from '%s': %s", client_id, scopes[0])
+            return False
+        return True
+
+    def validate_silent_authorization(self, _request) -> bool:
+        return True
+
+    def validate_silent_login(self, _request) -> bool:
+        return True
+
+    def validate_user_match(self, _id_token_hint, _scopes, _claims, _request) -> bool:
```

Comment:
Does this method override a method from the super class? The method parameter names should not be changed compared to the original super class since the parameter names are also part of the interface. Remember that even normal positional parameters can be supplied in the keyword argument style in function calls.

```python
# calling the method in the keyword style
myobj.validate_user_match(
    _id_token_hint=my_hint,
    _scopes=my_scopes,
    _claims=my_claims,
    _request=my_request,
)
```